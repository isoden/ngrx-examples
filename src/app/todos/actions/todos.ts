import { Action } from '@ngrx/store';
import { VisibilityFilter } from '../models';

export enum TodosActionTypes {
  Add                 = '[Todos] Add',
  Toggle              = '[Todos] Toggle',
  SetVisibilityFilter = '[Todos] Set Visibility Filter',
}

export class Add implements Action {
  readonly type = TodosActionTypes.Add;

  constructor(public payload: { title: string }) {}
}

export class Toggle implements Action {
  readonly type = TodosActionTypes.Toggle;

  constructor(public payload: { id: number }) {}
}

export class SetVisibilityFilter implements Action {
  readonly type = TodosActionTypes.SetVisibilityFilter;

  constructor(public payload: { visibilityFilter: VisibilityFilter }) {}
}

export type TodosActions =
  | Add
  | Toggle
  | SetVisibilityFilter;
