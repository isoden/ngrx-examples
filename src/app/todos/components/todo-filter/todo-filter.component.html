<p>
  Filter: 
  <button (click)="onFilterChange('SHOW_ALL')"       [disabled]="filter === 'SHOW_ALL'">All</button>
  <button (click)="onFilterChange('SHOW_ACTIVE')"    [disabled]="filter === 'SHOW_ACTIVE'">Active</button>
  <button (click)="onFilterChange('SHOW_COMPLETED')" [disabled]="filter === 'SHOW_COMPLETED'">Completed</button>
</p>
