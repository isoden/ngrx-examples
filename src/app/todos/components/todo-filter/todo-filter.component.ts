import { Component, OnInit, ChangeDetectionStrategy, Output, EventEmitter, Input } from '@angular/core';
import { VisibilityFilter } from '../../models';

@Component({
  selector: 'app-todo-filter',
  templateUrl: './todo-filter.component.html',
  styleUrls: ['./todo-filter.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TodoFilterComponent implements OnInit {
  @Input()
  filter: VisibilityFilter;

  @Output()
  filterChange = new EventEmitter<string>();

  constructor() { }

  ngOnInit() {
  }

  onFilterChange(visibilityFilter: VisibilityFilter) {
    this.filterChange.emit(visibilityFilter);
  }
}
