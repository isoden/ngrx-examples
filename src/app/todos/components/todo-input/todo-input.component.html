<input type="text" [formControl]="title">
<button (click)="submit()" type="button" [disabled]="title.invalid">Add todo</button>
