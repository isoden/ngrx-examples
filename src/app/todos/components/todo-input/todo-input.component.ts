import { Component, OnInit, ChangeDetectionStrategy, Output, EventEmitter } from '@angular/core';
import { FormControl, Validators } from '@angular/forms';

@Component({
  selector: 'app-todo-input',
  templateUrl: './todo-input.component.html',
  styleUrls: ['./todo-input.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TodoInputComponent implements OnInit {
  @Output()
  add = new EventEmitter<string>();

  title = new FormControl('', [
    Validators.required,
  ]);

  constructor() { }

  ngOnInit() {
  }

  submit() {
    if (this.title.invalid) {
      return;
    }

    this.add.emit(this.title.value);
    this.title.setValue('');
  }
}
