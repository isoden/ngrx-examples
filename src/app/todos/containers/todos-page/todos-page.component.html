<section>
  <h1 class="Heading -lv1">Todos</h1>

  <div class="Card">
    <app-todo-input
      (add)="onAdd($event)"
    ></app-todo-input>

    <app-todo
      *ngFor="let todo of todos$ | async"
      [todo]="todo"
      (toggle)="onToggle($event)"
    ></app-todo>

    <app-todo-filter
      [filter]="visibilityFilter$ | async"
      (filterChange)="onFilterChange($event)"
    ></app-todo-filter>
  </div>
</section>
