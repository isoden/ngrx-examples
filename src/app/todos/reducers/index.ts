import { createFeatureSelector, createSelector } from '@ngrx/store';
import * as fromItems from './items';
import * as fromVisibilityFilter from './visibility-filter';

export const reducers = {
  items           : fromItems.reducer,
  visibilityFilter: fromVisibilityFilter.reducer,
};

const {
  selectAll: selectAllTodos,
} = fromItems.adapter.getSelectors();

export interface TodosState {
  items: fromItems.TodosState;
  visibilityFilter: fromVisibilityFilter.VisibilityFilterState;
}

export interface State {
  todos: TodosState;
}

export const selectTodos = createFeatureSelector<TodosState>('todos');
export const selectTodosItems = createSelector(selectTodos, state => state.items);
export const selectTodosVisibilityFilter = createSelector(selectTodos, state => state.visibilityFilter);
export const getVisibilityFilter = createSelector(selectTodosVisibilityFilter, fromVisibilityFilter.getVisibilityFilter);
export const getTodos = createSelector(selectTodosItems, selectAllTodos);
export const getFilteredTodos = createSelector(getVisibilityFilter, getTodos, (filter, todos) => {
  if (filter === 'SHOW_ALL') {
    return todos;
  }

  return todos.filter(todo => {
    switch (filter) {
      case 'SHOW_ACTIVE': return !todo.completed;
      case 'SHOW_COMPLETED': return todo.completed;
    }
  });
});
