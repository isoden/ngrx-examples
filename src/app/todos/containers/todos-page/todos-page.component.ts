import { Component, OnInit, ChangeDetectionStrategy } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { combineLatest , of } from 'rxjs';
import { VisibilityFilter, Todo } from '../../models';
import * as fromTodos from '../../reducers';
import * as Todos from '../../actions/todos';

@Component({
  selector: 'app-todos-page',
  templateUrl: './todos-page.component.html',
  styleUrls: ['./todos-page.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TodosPageComponent implements OnInit {
  visibilityFilter$ = this.store.pipe(select(fromTodos.getVisibilityFilter));

  todos$ = combineLatest(
    this.store.pipe(select(fromTodos.getTodos)),
    this.visibilityFilter$,
    (todos, filter) => {
      const filterMap = {
        SHOW_ALL(todo: Todo) { return true; },
        SHOW_ACTIVE(todo: Todo) { return !todo.completed; },
        SHOW_COMPLETED(todo: Todo) { return todo.completed; },
      };

      return todos.filter(filterMap[filter]);
    }
  );

  constructor(
    private store: Store<fromTodos.State>,
  ) { }

  ngOnInit() {
  }

  onAdd(title: string) {
    this.store.dispatch(new Todos.Add({ title }));
  }

  onToggle(id: number) {
    this.store.dispatch(new Todos.Toggle({ id }));
  }

  onFilterChange(visibilityFilter: VisibilityFilter) {
    this.store.dispatch(new Todos.SetVisibilityFilter({ visibilityFilter }));
  }
}
