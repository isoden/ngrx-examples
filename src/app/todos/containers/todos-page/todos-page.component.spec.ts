import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { StoreModule, combineReducers, Store } from '@ngrx/store';
import { By } from '@angular/platform-browser';

import { TodosPageComponent } from './todos-page.component';
import { TodoComponent } from '../../components/todo/todo.component';
import { TodoFilterComponent } from '../../components/todo-filter/todo-filter.component';
import { TodoInputComponent } from '../../components/todo-input/todo-input.component';
import { reducers, State } from '../../reducers';
import { Add } from '../../actions/todos';

describe('TodosPageComponent', () => {
  let component: TodosPageComponent;
  let fixture: ComponentFixture<TodosPageComponent>;
  let store: Store<State>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        FormsModule,
        ReactiveFormsModule,
        StoreModule.forRoot({
          todos: combineReducers(reducers),
        }),
      ],
      declarations: [
        TodosPageComponent,
        TodoComponent,
        TodoFilterComponent,
        TodoInputComponent,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    store = TestBed.get(Store);
    fixture = TestBed.createComponent(TodosPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should compile', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch "Add"', () => {
    const title = 'Learn Angular';

    jest.spyOn(store, 'dispatch');
    component.onAdd(title);

    expect(store.dispatch).toHaveBeenCalledWith(new Add({ title }));
  });
});
