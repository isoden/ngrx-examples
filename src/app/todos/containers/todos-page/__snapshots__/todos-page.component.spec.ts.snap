// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosPageComponent should compile 1`] = `
<app-todos-page
  store={[Function Store]}
  todos$={[Function Observable]}
  visibilityFilter$={[Function Store]}
>
  <section>
    
    
    <h1
      class="Heading -lv1"
    >
      Todos
    </h1>
    
  
    
    <div
      class="Card"
    >
      
      
      <app-todo-input>
        <input
          class="ng-untouched ng-pristine ng-invalid"
          ng-reflect-form="[object Object]"
          type="text"
        />
        
  
        <button
          disabled=""
          type="button"
        >
          Add todo
        </button>
        
  
      </app-todo-input>
      
  
      
      
      
  
      
      <app-todo-filter
        ng-reflect-filter="SHOW_ALL"
      >
        <p>
          
    Filter: 
    
          <button
            disabled=""
          >
            All
          </button>
          
    
          <button>
            Active
          </button>
          
    
          <button>
            Completed
          </button>
          
  
        </p>
        
  
      </app-todo-filter>
      
    
    </div>
    
  
  </section>
  
</app-todos-page>
`;
